### PR TITLE
Schema registry integration

### LINKED ISSUE #13: AVRO Support
Is there any support for avro encoding / decoding via schema registry at the moment, I couldn't see anything in the configs.

### PR DESCRIPTION
**Pull request by drnushooz**
This is an integration for Nats Kafka bridge with Confluent compatible schema registries. The functionality uses srclient as the schema registry client and adds support for sending and receiving Avro, JSON and protobuf encoded messages. There are unit tests added as well.

Resolves #13

### CODE REVIEW

File: server/kafka/consumer_test.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+	schemaIDBytes := make([]byte, 4)
+	binary.BigEndian.PutUint32(schemaIDBytes, uint32(avroSchemaID))
+
+	var payload []byte
+	payload = append(payload, byte(0))
+	payload = append(payload, schemaIDBytes...)
+	payload = append(payload, avroBytes...)
+
+	_, avroError := (*saramaConsumer).deserializePayload(&consumer, payload)
```
Comment (variadico) [inline server/kafka/consumer_test.go:49]: I think we can simplify this.
```suggestion
	_, avroError := consumer.deserializePayload(payload)
```

File: server/kafka/mock_schema_registry.go [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+		"/schemas/ids/" + avroSchemaIDString:
+		avroSchemaType := srclient.Avro
+		responsePayload := schemaResponse{
+			Subject:    avroSubjectName,
+			Version:    avroSchemaVersion,
+			Schema:     avroSchema,
+			ID:         avroSchemaID,
+			SchemaType: &avroSchemaType,
+		}
+		response, _ := json.Marshal(responsePayload)
```
Comment (variadico) [inline server/kafka/mock_schema_registry.go:96]: Missing error check.

File: server/kafka/mock_schema_registry.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+	References []srclient.Reference `json:"references"`
+}
+
+type MockSchemaRegistry struct {
+	instance *httptest.Server
+	t        *testing.T
+}
+
+// NewMockSchemaServer Mock schema registry instance which is used to test the producer and consumer functionality
+func NewMockSchemaServer(t *testing.T) MockSchemaRegistry {
```
Comment (variadico) [inline server/kafka/mock_schema_registry.go:70]: Can we unexport `MockSchemaRegistry` and `NewMockSchemaServer`? Let's keep them private first. If needed, we can make them public later.

File: server/kafka/producer_test.go [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+	defer server.Close()
+
+	producer := saramaProducer{
+		schemaRegistryOn:     true,
+		schemaRegistryClient: srclient.CreateSchemaRegistryClient(server.GetServerURL()),
+		valueSubjectName:     avroSubjectName,
+		valueSchemaVersion:   avroSchemaVersion,
+	}
+
+	_, avroError := (*saramaProducer).serializePayload(&producer, []byte(avroMessage))
```
Comment (variadico) [inline server/kafka/producer_test.go:37]:
```suggestion
	_, avroError := producer.serializePayload([]byte(avroMessage))
```

**Review by variadico — COMMENTED** [review]
Just a few minor tweaks. Almost there!

**Comment by variadico** [comment]
Can we also add these new options to the docs here?
https://github.com/nats-io/nats-kafka/blob/main/docs/config.md?plain=1#L243-L255

File: server/kafka/mock_schema_registry.go (hunk 2 above)
Comment (drnushooz) [reply server/kafka/mock_schema_registry.go]: Done.

File: server/kafka/mock_schema_registry.go (hunk 3 above)
Comment (drnushooz) [reply server/kafka/mock_schema_registry.go]: Done.

File: server/kafka/consumer_test.go (hunk 1 above)
Comment (drnushooz) [reply server/kafka/consumer_test.go]: Done.

File: server/kafka/consumer_test.go [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+	defer server.Close()
+
+	consumer := saramaConsumer{
+		schemaRegistryOn:     true,
+		schemaRegistryClient: srclient.CreateSchemaRegistryClient(server.GetServerURL()),
+	}
+
+	avroCodec, _ := goavro.NewCodec(avroSchema)
+	native, _, _ := avroCodec.NativeFromTextual([]byte(avroMessage))
+	avroBytes, _ := avroCodec.BinaryFromNative(nil, native)
```
[the remark this answers is not in the document]
Comment (drnushooz) [reply server/kafka/consumer_test.go:47]: Yep updated them in the latest revision.

**Comment by drnushooz** [comment]
> 

The documentation has been updated

File: server/kafka/consumer.go [hunk 6]
```diff
@@ -261,3 +309,68 @@ func (c *saramaConsumer) ConsumeClaim(sess sarama.ConsumerGroupSession, claim sa
 
 	return nil
 }
+
+// Retrieve the schema of the message and deserialize it.
+func (c *saramaConsumer) deserializePayload(payload []byte) ([]byte, error) {
+	// first byte of the payload is 0
+	if payload[0] != byte(0) {
+		return nil, fmt.Errorf("magic byte is not 0")
```
Comment (variadico) [inline server/kafka/consumer.go:317]: Can we change this to something like "failed to deserialize payload: magic byte is not 0"?

File: server/kafka/mock_schema_registry.go [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+// schemas for testing
+const avroSchema = "{" +
+	"\"type\": \"record\"," +
+	"\"name\": \"snack\"," +
+	"\"fields\": [" +
+	"{\"name\": \"name\", \"type\": \"string\"}," +
+	"{\"name\": \"manufacturer\", \"type\": \"string\"}," +
+	"{\"name\": \"calories\", \"type\": \"float\"}," +
+	"{\"name\": \"color\", \"type\": [\"null\", \"string\"], \"default\": null}" +
+	"]}"
```
Comment (variadico) [inline server/kafka/mock_schema_registry.go:40]: We can use raw string literals (with backticks) here to make this a little more readable and writable.
```go
const avroSchema = `{
	"type": "record",
	"name": "snack",
	"fields": [
	{"name": "name", "type": "string"},
	{"name": "manufacturer", "type": "string"},
	{"name": "calories", "type": "float"},
	{"name": "color", "type": ["null", "string"], "default": null}
	]}`
```

**Review by variadico — APPROVED** [review]
I just have 2 final tweaks we can make. Otherwise, LGTM! We can merge today!

File: server/kafka/consumer.go (hunk 6 above)
Comment (drnushooz) [reply server/kafka/consumer.go:317]: Done.

File: server/kafka/mock_schema_registry.go (hunk 7 above)
Comment (drnushooz) [reply server/kafka/mock_schema_registry.go]: Done.

**Comment by variadico** [comment]
Awesome, thank you very much! 🙏🏽  A new release will be coming soon!

**Comment by drnushooz** [comment]
> Awesome, thank you very much! 🙏🏽 A new release will be coming soon!

It's been a pleasure. Looking forward to the next release :).

**Pull request merged**